src/IIIFPresentation/BackgroundHandler/BatchCompletion/BatchCompletionMessageHandler.cs

```diff
@@ -0,0 +1,162 @@
+using Models.Database.General;
+using Models.DLCS;
+using Repository;
+using Batch = Models.Database.General.Batch;
+
+namespace BackgroundHandler.BatchCompletion;
+
+public class BatchCompletionMessageHandler(
+    PresentationContext dbContext,
+    IDlcsOrchestratorClient dlcsOrchestratorClient,
+    IOptions<DlcsSettings> dlcsOptions,
+    IIIIFS3Service iiifS3,
+    ILogger<BatchCompletionMessageHandler> logger)
+    : IMessageHandler
+{
+    private static readonly JsonSerializerOptions JsonSerializerOptions = new(JsonSerializerDefaults.Web);
+    private readonly DlcsSettings dlcsSettings = dlcsOptions.Value;
+
+    public async Task<bool> HandleMessage(QueueMessage message, CancellationToken cancellationToken)
+    {
+        using (LogContextHelpers.SetServiceName(nameof(BatchCompletionMessageHandler)))
+        {
+            try
+            {
+                var batchCompletionMessage = DeserializeMessage(message);
+
+                await UpdateAssetsIfRequired(batchCompletionMessage, cancellationToken);
+                return true;
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(ex, "Error handling batch-completion message {MessageId}", message.MessageId);
+            }
+        }
+        
+        return false;
+    }
+
+    private async Task UpdateAssetsIfRequired(BatchCompletionMessage batchCompletionMessage, CancellationToken cancellationToken)
+    {
+        var batch = await dbContext.Batches.Include(b => b.Manifest)
+            .ThenInclude(m => m.CanvasPaintings)
+            .SingleOrDefaultAsync(b => b.Id == batchCompletionMessage.Id, cancellationToken);
+        
+        // batch isn't tracked by presentation, so nothing to do
+        if (batch == null) return;
+        
+        // Other batches haven't completed, so no point populating items until all are complete
+        if (await dbContext.Batches.AnyAsync(b => b.ManifestId == batch.ManifestId &&
+                                                  b.Status != BatchStatus.Completed &&
+                                                  b.Id != batch.Id, cancellationToken))
+        {
+            CompleteBatch(batch, batchCompletionMessage.Finished);
+        }
+        else
+        {
+            logger.LogInformation(
+                "Attempting to complete assets in batch {BatchId} for customer {CustomerId} with the manifest {ManifestId}",
+                batch.Id, batch.CustomerId, batch.ManifestId);
+
+            var batches = dbContext.Batches.Where(b => b.ManifestId == batch.ManifestId).Select(b => b.Id).ToList();
+
+            var generatedManifest =
+                await dlcsOrchestratorClient.RetrieveAssetsForManifest(batch.CustomerId, batches,
+                    cancellationToken);
+            
+            Dictionary<AssetId, Canvas> itemDictionary;
+            
+            try
+            {
+                itemDictionary = generatedManifest.Items.Select(i =>
```

Comment:
As discussed, make canvas id parsing a method to ease of testing and changing. Also, `.ToDictionary()` would save casting to KVP and increase readability